### PR TITLE
DLPack: add support to PyTorch/XLA

### PR DESCRIPTION
**Pull request by ysiraichi**
Taking over: #128176.

In summary, this PR:

- `__dlpack__`: Calls PyTorch/XLA `to_dlpack` function, if the tensor lives in an XLA:CUDA device
- `__dlpack_device__`: Correctly maps PyTorch/XLA tensors to `kDLGPU`, if XLA:CUDA is being used

The tests are introduced in https://github.com/pytorch/xla/pull/7213.

### CODE REVIEW

**Comment by ysiraichi** [comment]
@albanD @JackCaoG Could you review this PR whenever you got time? This is basically #128176 + rebase + fix. Let me know if there are any changes you'd like to see.

File: torch/_tensor.py [hunk 1]
```diff
@@ -1616,6 +1616,16 @@ def __dlpack__(self, stream=None):
...
                     stream.wait_event(event)
+        if self.device.type == "xla":
+            import torch_xla
+            import torch_xla.utils.dlpack as xla_dlpack
+
+            if (
+                len(torch_xla.real_devices()) <= 0
+                or "cuda" not in torch_xla.real_devices()[0].lower()
+            ):
+                raise RuntimeError("Can't export XLA tensors not on CUDA.")
```
Comment (albanD) [inline torch/_tensor.py:1627]:
```suggestion
                raise RuntimeError("Can't export to dlpack an XLA tensors that is not on CUDA.")
```

**Review by albanD — APPROVED** [review]
SGTM

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by albanD** [comment]
Please fix the lint!

**Pull request closed without merging**